src/Exception/DuplicateKeyException.php

```diff
@@ -0,0 +1,37 @@
+<?php
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Exception;
+
+use MongoDB\Driver\Exception\BulkWriteException;
+
+class DuplicateKeyException extends BulkWriteException implements Exception
+{
+    /**
+     * Wraps a BulkWriteException.
+     *
+     * @param BulkWriteException $e
+     * @return self
+     */
+    public static function fromBulkWriteException(BulkWriteException $e)
+    {
+        $duplicateKeyException = new static($e->getMessage(), $e->getCode(), $e);
+        $duplicateKeyException->writeResult = $e->getWriteResult();
```

Comment:
Unfortunately, PHP won't let us cast exceptions, so we have to resort to chaining (passing the previous exception in our new exception's constructor). The message, code, and WriteResult must be manually copied. [This thread on Stack Overflow](https://stackoverflow.com/a/5551828/162228) was helpful to understand some relevant best practices.

By sheer coincidence, PHPC tracks the WriteResult object as a protected property (as opposed to the usual way of zval pointers without our structs).